lib/aws/client.ex

```diff
@@ -10,7 +10,15 @@ defmodule AWS.Client do
             service: nil,
             endpoint: "amazonaws.com",
             proto: "https",
-            port: "443"
+            port: "443",
+            http_client: {HTTPoison, :request},
+            encode: %{json: {Poison.Parser, :parse!},
+                      xml: {AWS.Util, :encode_xml},
+                      query: {AWS.Util, :encode_query}},
+            decode: %{json: {Poison.Encoder, :encode},
+                      xml: {AWS.Util, :decode_xml},
+                      query: {AWS.Util, :decode_xml}}
```

Comment:
Those representations are fine for now but I think our APIs should be:

    json_module: {Foo, options},
    xml_module: {Bar, options}
    http_module: {Baz, options}

The reason I don't like the map is because there is no easy to change just one of them (I either need to set the whole map or use put_in or similar for a nested update). I think the module approach with behaviour definition is better but we can do that separately. WDYT? Would you prefer to tackle it now or later?

Reply:
And of course, no module for the query for now. :) We can pattern match it during encode/decode though and handle it locally.

Reply:
https://github.com/aws-beam/aws-elixir/pull/42#discussion_r514615261

Let's do it now. 😄

Reply:
Fantastic @vrcca! Btw, if you are looking for more ways to contribute, we would love help with https://github.com/aws-beam/aws-codegen/issues/54 too. The main reason is that it will make it easier for us to see exactly which modules we have to implement/provide (and this PR will add some) and which ones are generated.

Thank you for all the contributions so far! <3

Reply:
I fail to see what the `options` for each `format_module` would be. Would you care to show an example? I've seen some libraries that allow to pass implementations like this:

```elixir
{Mod, Fun, Opts}
{Mod} | Mod
{Mod, Fun}
fn input -> ... end
```

Does it makes sense to add this flexibility, or just accept the `Mod` and assume that they all implement `encode!` & `decode!`?

I like passing {Mod, Fun} because it's easy to reuse libs without the need to create a module just for this.

Reply:
The options can be useful so we provide default adapters and allow people to customize them. For example, if we are using hackney, I could depend on aws_elixi_hackney and do this:

    http_client: {AWS.Elixir.Hackney, pool: :my_pool_that_aws_elixir_does_not_know_about}

> Does it makes sense to add this flexibility, or just accept the Mod and assume that they all implement encode! & decode!?

That would be my route, yes. I know the module can be annoying but, as we add more functions, it will provide more structure, better place to put our docs, etc.

Reply:
I get it now. So maybe we should get rid of the `options` param from all generated methods. I believe this parameter exists only for the HTTP client.

Reply:
@vrcca I am not sure. For example, for Jason, you may want to control in some cases if it should copy binaries or create references as it decodes. :) so I think it is generally applicable and it can be a nice escape hatch, otherwise people will have to define modules for each possible option they may want.